crates/iota-types/src/dynamic_field/visitor.rs

```diff
@@ -0,0 +1,533 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::{
+    account_address::AccountAddress,
+    annotated_value as A,
+    annotated_visitor::{self, StructDriver, ValueDriver, VariantDriver, VecDriver, Visitor},
+    language_storage::TypeTag,
+    u256::U256,
+};
+
+use super::{DynamicFieldInfo, DynamicFieldType};
+use crate::{base_types::ObjectID, id::UID};
+
+/// Visitor to deserialize the outer structure of a `0x2::dynamic_field::Field`
+/// while leaving its name and value untouched.
+pub struct FieldVisitor;
+
+#[derive(Debug, Clone)]
+pub struct Field<'b, 'l> {
+    pub id: ObjectID,
+    pub kind: DynamicFieldType,
+    pub name_layout: &'l A::MoveTypeLayout,
+    pub name_bytes: &'b [u8],
+    pub value_layout: &'l A::MoveTypeLayout,
+    pub value_bytes: &'b [u8],
+}
+
+pub enum ValueMetadata {
+    DynamicField(TypeTag),
+    DynamicObjectField(ObjectID),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Not a dynamic field")]
+    NotADynamicField,
+
+    #[error("Not a dynamic object field")]
+    NotADynamicObjectField,
+
+    #[error("{0}")]
+    Visitor(#[from] annotated_visitor::Error),
+}
+
+impl FieldVisitor {
+    /// Deserialize the top-level structure from a dynamic field's
+    /// `0x2::dynamic_field::Field` without having to fully deserialize its
+    /// name or value.
+    pub fn deserialize<'b, 'l>(
+        bytes: &'b [u8],
+        layout: &'l A::MoveTypeLayout,
+    ) -> anyhow::Result<Field<'b, 'l>> {
+        A::MoveValue::visit_deserialize(bytes, layout, &mut FieldVisitor)
+    }
+}
+
+impl Field<'_, '_> {
```

Comment:
Note that here we use `Field<'_, '_>` instead of using `impl<'b, 'l> Field<'b, 'l>` in the upstream change to fix the Lint error.